cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+
+```
+granularity = 5
+min_extra_capacity_rate = 0.1
+cluster_cpu_capacity = sum(node_cpu_capacity)
+cluster_mem_capacity = sum(node_mem_capacity)
+extra_cpu = cluster_cpu_capacity * min_extra_capacity_rate
+extra_mem = cluster_mem_capacity * min_extra_capacity_rate
+biggest_cpu_pod_idx = idx where max(pod_requested_cpu(idx)) 
+biggest_cpu_pod_cpu = pod_requested_cpu(biggest_cpu_pod_idx)
+biggest_cpu_pod_mem = pod_requested_mem(biggest_cpu_pod_idx)
+biggest_mem_pod_idx = idx where max(pod_requested_mem(idx))
+biggest_mem_pod_cpu = pod_requested_cpu(biggest_mem_pod_idx)
+biggest_mem_pod_mem = pod_requested_mem(biggest_mem_pod_idx)
+remaining_cpu = extra_cpu - biggest_cpu_pod_cpu - biggest_mem_pod_cpu
+remaining_mem = extra_mem - biggest_cpu_pod_mem - biggest_mem_pod_mem
+placeholder_cpu = remaining_cpu / placeholder_cnt
+placeholder_mem = remaining_mem / placeholder_cnt
+placeholder_cnt = cluster_size * granularity
+```
+
+With that there will be an always present in-cluster-autoscaler need for
+
+```
+biggest_cpu_pod_cpu + biggest_mem_pod_cpu + placeholder_cpu * placeholder_cnt
+= biggest_cpu_pod_cpu + biggest_mem_pod_cpu + remaining_cpu / placeholder_cnt * placeholder_cnt
+= biggest_cpu_pod_cpu + biggest_mem_pod_cpu + remaining_cpu
+= extra_cpu
+```
+
+cpu and
+
+```
+biggest_cpu_pod_mem + biggest_mem_pod_mem + placeholder_mem * placeholder_cnt
+= biggest_cpu_pod_mem + biggest_mem_pod_mem + remaining_mem / placeholder_cnt * placeholder_cnt
+= biggest_cpu_pod_mem + biggest_mem_pod_mem + remaining_mem
+= extra_mem
+```
+
+#### In version 2 (More complete idea)
+
+Have `N` placeholder pods, each requiring `placeholder_cpu` amount of cpu and `placeholder_mem` amount of memory. The variables would be defined as:
+
+Limitation: This produces a minimum possible extra space([R5]) while producing enough space for any pod of any size([R6], [R7])
+
+```
+granularity = 5
+min_extra_capacity_rate = 0.1
+cluster_cpu_capacity = sum(node_cpu_capacity)
+cluster_mem_capacity = sum(node_mem_capacity)
+extra_cpu = cluster_cpu_capacity * min_extra_capacity_rate
+extra_mem = cluster_mem_capacity * min_extra_capacity_rate
+max_pod_cpu = max(pod_requested_cpu)
+max_pod_mem = max(pod_requested_mem)
+remaining_cpu = extra_cpu - max_pod_cpu
+remaining_mem = extra_mem - max_pod_mem
+placeholder_cpu = remaining_cpu / placeholder_cnt
+placeholder_mem = remaining_mem / placeholder_cnt
+placeholder_cnt = cluster_size * granularity
+```
+
+With that there will be an always present in-cluster-autoscaler need for
+
+```
+max_pod_cpu + placeholder_cpu * placeholder_cnt
+= max_pod_cpu + remaining_cpu / placeholder_cnt * placeholder_cnt
+= max_pod_cpu + remaining_cpu
+= extra_cpu
+```
+
+cpu and
+
+```
+max_pod_mem + placeholder_mem * placeholder_cnt
+= max_pod_mem + remaining_mem / placeholder_cnt * placeholder_cnt
+= max_pod_mem + remaining_mem
+= extra_mem
+```
+
+### How To Test
+
+#### E2E
+
+* Provide a document to reproduce the test scenario
```

Comment:
We have e2e test that already do most of what is described here (provision cluster, etc). Ideally they should be modified to run on AWS, but that's a separate issue not related to this.

Also we're testing cluster-autoscaler not HPA, so I'd rather have the test just add more pods, rather than generate load and rely on HPA to do this (reduce number of moving parts to reduce flakiness).